onnxruntime/core/providers/cpu/ml/treeregressor.cc

```diff
@@ -79,149 +60,216 @@ TreeEnsembleRegressor<T>::TreeEnsembleRegressor(const OpKernelInfo& info)
-
-    return std::get<1>(t1) < std::get<1>(t2);
-  });
-  //make an index so we can find the leafnode data quickly when evaluating
-  int64_t field0 = -1;
-  int64_t field1 = -1;
-  for (size_t i = 0; i < leafnode_data_.size(); i++) {
-    int64_t id0 = std::get<0>(leafnode_data_[i]);
-    int64_t id1 = std::get<1>(leafnode_data_[i]);
-    if (id0 != field0 || id1 != field1) {
-      int64_t id = id0 * four_billion_ + id1;
-      auto p3 = std::make_pair(id, i);  // position is i
-      leafdata_map_.insert(p3);
-      field0 = id;
-      field1 = static_cast<int64_t>(i);
-    }
-  }
-  //treenode ids, some are roots, and roots have no parents
-  std::unordered_map<int64_t, size_t> parents;  //holds count of all who point to you
-  std::unordered_map<int64_t, size_t> indices;
-  //add all the nodes to a map, and the ones that have parents are not roots
-  std::unordered_map<int64_t, size_t>::iterator it;
-  size_t start_counter = 0L;
-  for (size_t i = 0; i < nodes_treeids_.size(); i++) {
-    //make an index to look up later
-    int64_t id = nodes_treeids_[i] * four_billion_ + nodes_nodeids_[i];
-    auto p3 = std::make_pair(id, i);  // i is the position
-    indices.insert(p3);
-    it = parents.find(id);
-    if (it == parents.end()) {
-      //start counter at 0
-      auto p1 = std::make_pair(id, start_counter);
-      parents.insert(p1);
-    }
+
+  // Filling nodes_.
+  nbnodes_ = nodes_treeids_.size();
+  nodes_ = new TreeNodeElement[(int)nbnodes_];
+  roots_.clear();
+  std::map<TreeNodeElementId, TreeNodeElement*> idi;
+  size_t i;
+
+  for (i = 0; i < nodes_treeids_.size(); ++i) {
+    TreeNodeElement* node = nodes_ + i;
+    node->id.tree_id = (int)nodes_treeids_[i];
+    node->id.node_id = (int)nodes_nodeids_[i];
+    node->feature_id = (int)nodes_featureids_[i];
+    node->value = nodes_values_[i];
+    node->hitrates = i < nodes_hitrates_.size() ? nodes_hitrates_[i] : -1;
+    node->mode = nodes_modes_[i];
+    node->truenode = NULL;   // nodes_truenodeids_[i];
+    node->falsenode = NULL;  // nodes_falsenodeids_[i];
+    node->missing_tracks = i < (size_t)missing_tracks_true_.size()
+                               ? (missing_tracks_true_[i] == 1
+                                      ? MissingTrack::TRUE
+                                      : MissingTrack::FALSE)
+                               : MissingTrack::NONE;
+    ORT_ENFORCE(idi.find(node->id) == idi.end());
+    idi.insert(std::pair<TreeNodeElementId, TreeNodeElement*>(node->id, node));
   }
-  //all true nodes aren't roots
-  for (size_t i = 0; i < nodes_truenodeids_.size(); i++) {
-    if (nodes_modes_[i] == ::onnxruntime::ml::NODE_MODE::LEAF) continue;
-    //they must be in the same tree
-    int64_t id = nodes_treeids_[i] * offset_ + nodes_truenodeids_[i];
-    it = parents.find(id);
-    ORT_ENFORCE(it != parents.end());
-    it->second++;
+
+  TreeNodeElementId coor;
+  TreeNodeElement* it;
+  for (i = 0; i < (size_t)nbnodes_; ++i) {
+    it = nodes_ + i;
```

Comment:
Could nbnodes_ be size_t instead of int64? It's initialized from a size_t and cast base to size_t multiple times.